library/src/main/java/com/datatorrent/lib/join/AbstractJoinOperator.java

```diff
@@ -0,0 +1,343 @@
+ * applies the join operation based on constraint and emit the joined value.
+ * Subclasses should provide implementation to createOutputTuple,copyValue, getKeyValue, getTime methods.
+ *
+ * <b>Properties:</b><br>
+ * <b>expiryTime</b>: Expiry time for stored tuples<br>
+ * <b>includeFieldStr</b>: List of comma separated fields to be added to the output tuple.
+ *                         Ex: Field1,Field2;Field3,Field4<br>
+ * <b>keyFields</b>: List of comma separated key field for both the streams. Ex: Field1,Field2<br>
+ * <b>timeFields</b>: List of comma separated time field for both the streams. Ex: Field1,Field2<br>
+ * <b>bucketSpanInMillis</b>: Span of each bucket in milliseconds.<br>
+ * <b>strategy</b>: Type of join operation. Default type is inner join<br>
+ * <br>
+ * </p>
+ *
+ * @displayName Abstract Join Operator
+ * @tags join
+ */
+@Evolving
+public abstract class AbstractJoinOperator<T> extends BaseOperator implements Operator.CheckpointListener
+{
+  public final transient DefaultOutputPort<List<T>> outputPort = new DefaultOutputPort<>();
+  protected String[] timeFields;
+
+  protected String[][] includeFields;
+  // Fields to compare from both the streams
+  protected String[] keys;
+  // Strategy of Join operation, by default the option is inner join
+  protected JoinStrategy strategy = JoinStrategy.INNER_JOIN;
+  // This represents whether the processing tuple is from left port or not
+  protected Boolean isLeft;
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<T> input1 = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      isLeft = true;
+      processTuple(tuple);
+    }
+  };
+  @InputPortFieldAnnotation(optional = true)
+  public transient DefaultInputPort<T> input2 = new DefaultInputPort<T>()
+  {
+    @Override
+    public void process(T tuple)
+    {
+      isLeft = false;
+      processTuple(tuple);
+    }
+  };
+  // Stores for each of the input port
+  private BackupStore[] store = (BackupStore[])Array.newInstance(BackupStore.class, 2);
+  private String includeFieldStr;
+  private String keyFieldStr;
+  private String timeFieldStr;
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    if (store[0] == null) {
+      throw new RuntimeException("Left Store is Empty");
+    }
+    if (store[1] == null) {
+      throw new RuntimeException("Right Store is Empty");
+    }
+    // Checks whether the strategy is outer join and set it to store
+    boolean isOuter = strategy.equals(JoinStrategy.LEFT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN);
+    store[0].isOuterJoin(isOuter);
+    isOuter = strategy.equals(JoinStrategy.RIGHT_OUTER_JOIN) || strategy.equals(JoinStrategy.OUTER_JOIN);
+    store[1].isOuterJoin(isOuter);
+    // Setup the stores
+    store[0].setup();
+    store[1].setup();
+
+    populateFields();
```

Comment:
This is called every time setup is called then why are timeFields[], keys[], includeFields[] defined as non-transient